[Issues 2403][xray] Fix raylet performance issues on scheduling queue

zhijunfu opened the pull request:
NOTE:    As mentioned in the issue thread, there are some other PRs related to scheduling queue, current plan is to wait for these PRs to merge first, and revisit this PR. 

## What do these changes do?

During performance testing, I found the latest raylet has significant performance regressions when I tried to transfer 100K integers from one actor to another using a python script, on my local macbook. The script previously takes 40s but right now cannot finish in 15+ minutes.

With the fix, running the same script on raylet gets back down to about 40s again.

## Related issue number

#2403

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6731/
Test FAILed.

raulchen commented on src/ray/raylet/scheduling_queue.h:
```diff
@@ -116,21 +116,35 @@ class SchedulingQueue {
+    bool Remove(const TaskID &task_id);
+    bool Remove(const TaskID &task_id, std::vector<Task> &removed_tasks);
+
+    const std::list<Task> &GetTasks() const;
```
nit: better add some simple documents for these functions.

zhijunfu replied:
Accepted. thanks!

atumanov commented on src/ray/raylet/scheduling_queue.h:
```diff
@@ -116,21 +116,35 @@ class SchedulingQueue {
+
+    const std::list<Task> &GetTasks() const;
+  private:
+    std::list<Task> list_;
```
let's use more descriptive names, perhaps a `task_list_` and `task_map_`?

zhijunfu replied:
makes sense.

atumanov commented on src/ray/raylet/scheduling_queue.h:
```diff
@@ -116,21 +116,35 @@ class SchedulingQueue {
+  public:
+    TaskQueue() {}
+    bool Append(const TaskID &task_id, const Task &task);
+    bool Remove(const TaskID &task_id);
```
I would change the name to  `RemoveTask` and `RemoveTasks` to make the difference explicit. Also, do we need `RemoveTask` for a single task if it's just a special case of `RemoveTasks`?

zhijunfu replied:
will rename to ```AppendTask()``` and ```RemoveTask()``` respectively.  For the two ```Remove``` functions, they are actually both removing a single task, it's just one of them will append the removed task to the passed in vector if removal succeeds.

atumanov commented on src/ray/raylet/scheduling_queue.h:
```diff
@@ -116,21 +116,35 @@ class SchedulingQueue {
+ 
   /// Tasks that are destined for actors that have not yet been created.
-  std::list<Task> uncreated_actor_methods_;
+  TaskQueue uncreated_actor_methods_;
```
yes, abstracting the data structure used to store tasks makes sense.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -6,49 +6,90 @@ namespace ray {
-inline void queueTasks(std::list<Task> &queue, const std::vector<Task> &tasks) {
-  queue.insert(queue.end(), tasks.begin(), tasks.end());
+inline void queueTasks(SchedulingQueue::TaskQueue &queue, const std::vector<Task> &tasks) {
+  for (auto &task : tasks) {
```
how about `const auto &task : tasks`

zhijunfu replied:
that's better:)

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -6,49 +6,90 @@ namespace ray {
-  queue.insert(queue.end(), tasks.begin(), tasks.end());
+inline void queueTasks(SchedulingQueue::TaskQueue &queue, const std::vector<Task> &tasks) {
+  for (auto &task : tasks) {
+    queue.Append(task.GetTaskSpecification().TaskId(), task);
```
hmm, might be worth adding `AppendTasks` as one of the public methods of the `TaskQueue` class. Both the unordered map and list have `insert()` that can take two iterators pointing to the beginning and the end of the range of elements to insert. This will make it possible to avoid calling Append in a for loop.

zhijunfu replied:
Using ```insert``` to add a range of elements to ```list``` is fine. The problem is how to add these to ```unordered_map```,  which contains pairs of ```TaskID, list::iterator``` - to do that we need to create a temporary vector of these pairs, in order to call ```insert``` on ```unordered_map```.  I feel like it's probably better to keep the current approach using ```AppendTask```.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -6,49 +6,90 @@ namespace ray {
 
 namespace raylet {
 
+bool SchedulingQueue::TaskQueue::Append(const TaskID &task_id, const Task &task) {
```
general style comment: we try to make all variable names use full words (concatenated with an _ ) and, to the extent possible, we strive to make those variable names as descriptive as possible. Examples `list_iter`, `list_`, `map_`

zhijunfu replied:
Got it. Thanks for the comment!:)

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -6,49 +6,90 @@ namespace ray {
 
+bool SchedulingQueue::TaskQueue::Append(const TaskID &task_id, const Task &task) {
+  if (map_.find(task_id) != map_.end()) {
+    // This shouldn't happen.
```
if this shouldn't happen, we need to `RAY_CHECK` and fail hard if it does happen. Basically, any time you think something shouldn't happen -- it's an assumption that we must have an assert for.

zhijunfu replied:
revised. thanks.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -6,49 +6,90 @@ namespace ray {
+bool SchedulingQueue::TaskQueue::Remove(const TaskID &task_id) {
+  auto iter = map_.find(task_id);
+  if (iter == map_.end()) {
+    // not exist.
```
We try to make all comments be complete, capitalized sentences. E.g., "Specified task was not found in the map". In this particular case this is also obvious from the code, so this comment is unnecessary.

zhijunfu replied:
sure.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -6,49 +6,90 @@ namespace ray {
+bool SchedulingQueue::TaskQueue::Remove(const TaskID &task_id, std::vector<Task> &removed_tasks) {
+  auto iter = map_.find(task_id);
+  if (iter == map_.end()) {
+    // not exist.
```
This comment can be removed. The code is self-explanatory in this case.

zhijunfu replied:
ok.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -6,49 +6,90 @@ namespace ray {
+
+  auto list_iter = iter->second;
+  removed_tasks.push_back(std::move(*list_iter));
+  map_.erase(iter);
```
is it safe to erase after std::move on this element (e.g., should we be worried about a double free here)?

zhijunfu replied:
It should be safe here. After ```std::move``` the specific task object as specified by ```*list_iter``` is not referenced any more, except for destruction, which is safe.  The ```erase``` basically removes the list iterator from hash, which is not impacted by std::move.

atumanov commented on src/ray/raylet/scheduling_queue.h:
```diff
@@ -116,21 +116,35 @@ class SchedulingQueue {
 
+  class TaskQueue {
+  public:
+    TaskQueue() {}
```
need to have a destructor, which flushes all elements from the task list and the task hashmap. Otherwise we might have a memory leak here.

zhijunfu replied:
thanks, revised.

zhijunfu commented:
@atumanov Many thanks for the detailed review. I've rebased to latest, revised code and addressed the comments. May you please help to take another look?

A small update: I added a ```HasTask()``` to ```TaskQueue``` to optimize the new added ```FilterStateFromQueue``` function.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6949/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6956/
Test FAILed.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -45,28 +45,72 @@ namespace ray {
+
+bool SchedulingQueue::TaskQueue::AppendTask(const TaskID &task_id, const Task &task) {
+  RAY_CHECK(task_map_.find(task_id) == task_map_.end());
+  auto list_iter = task_list_.insert(task_list_.end(), task);
```
I know it's annoying, but the style policy for Ray is to have full, descriptive variable names. In this case `list_iterator` and `iterator` instead of `list_iter` and `iter`.

zhijunfu replied:
update. thanks.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -45,28 +45,72 @@ namespace ray {
+}
+
+bool SchedulingQueue::TaskQueue::RemoveTask(const TaskID &task_id) {
+  auto iter = task_map_.find(task_id);
```
instead of `iter` how about `task_found_iterator`

zhijunfu replied:
revised.

atumanov commented on src/ray/raylet/scheduling_queue.cc:
```diff
@@ -45,28 +45,72 @@ namespace ray {
+
+bool SchedulingQueue::TaskQueue::RemoveTask(const TaskID &task_id,
+                                            std::vector<Task> &removed_tasks) {
+  auto iter = task_map_.find(task_id);
```
`task_found_iterator`

zhijunfu replied:
ok.

atumanov commented on src/ray/raylet/scheduling_queue.h:
```diff
@@ -151,21 +151,62 @@ class SchedulingQueue {
+    ///
+    /// \param task_id The task ID for the task to remove from the queue.
+    /// \param removed_tasks If the task specified by task_id is successfully
+    //  removed from the queue, the task data is appened to the vector..
```
sp: *appended* to the vector..

zhijunfu replied:
thanks for catching this.

atumanov commented on src/ray/raylet/scheduling_queue.h:
```diff
@@ -151,21 +151,62 @@ class SchedulingQueue {
+    /// \brief Append a task to queue.
+    ///
+    /// \param task_id The task ID for the task to append.
+    /// \param task The task to append to the queue.
```
for all methods with a return value, please also document the return value using `\return`

atumanov commented:
You may also add me as a collaborator on `ant-tech-alliance` so that I could push small fixes directly to your branch, if this helps.

zhijunfu commented:
@atumanov Thanks for reviewing. Rebased to latest master and addressed comments. 
btw I don't have permission to add ```ant-tech-alliance``` collaborators, let me ask and see if it's doable.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7093/
Test PASSed.

atumanov commented:
changes look good. Currently looking into some test failures on Travis:
```
=================================== FAILURES ===================================
[31m[1m______________________ ResourcesTest.testCustomResources _______________________[0m

self = <runtest.ResourcesTest testMethod=testCustomResources>

[1m    def testCustomResources(self):[0m
[1m        ray.worker._init([0m
[1m            start_ray_local=True,[0m
[1m            num_local_schedulers=2,[0m
[1m            num_cpus=[3, 3],[0m
[1m            resources=[{[0m
[1m                "CustomResource": 0[0m
[1m            }, {[0m
[1m                "CustomResource": 1[0m
[1m            }])[0m
[1m    [0m
[1m        @ray.remote[0m
[1m        def f():[0m
[1m            time.sleep(0.001)[0m
[1m            return ray.worker.global_worker.plasma_client.store_socket_name[0m
[1m    [0m
[1m        @ray.remote(resources={"CustomResource": 1})[0m
[1m        def g():[0m
[1m            time.sleep(0.001)[0m
[1m            return ray.worker.global_worker.plasma_client.store_socket_name[0m
[1m    [0m
[1m        @ray.remote(resources={"CustomResource": 1})[0m
[1m        def h():[0m
[1m            ray.get([f.remote() for _ in range(5)])[0m
[1m            return ray.worker.global_worker.plasma_client.store_socket_name[0m
[1m    [0m
[1m        # The f tasks should be scheduled on both local schedulers.[0m
[1m>       self.assertEqual(len(set(ray.get([f.remote() for _ in range(50)]))), 2)[0m
[1m[31mE       AssertionError: 1 != 2[0m

[1m[31mtest/runtest.py[0m:1816: AssertionError
```
This particular test failure appears unrelated.
